h2o-core/src/main/java/water/api/ParseHandler.java

```diff
@@ -51,6 +63,36 @@ public ParseV3 parse(int version, ParseV3 parse) {
     return parse;
   }
 
+  /**
+   * Extracts synthetic column values from the keys of parsed files, as the keys contain path to the file and the
+   * partitioned file path contains the values necessary for all the columns the dataset is partitioned by.
+   *
+   * @param partitionColumnNames Names of the columns to be partitioned. Those are expected to be validated in
+   *                             the ParseSetup phase.
+   * @param fileKeys             Keys to all the files parsed, with key IDs having the file path in them.
+   * @return A two-dimensional {@link String} array, where the first dimension is the index of the column equal to the
+   * partitionColumnNames. The second dimension are the categorical values the datasaet has been partitioned by.
+   * @throws IllegalArgumentException If one of the partitioned columns is not found in any of the paths provided.
+   */
+  private static String[][] syntheticColumValuesFromPartitions(final String[] partitionColumnNames, final Key[] fileKeys)
+          throws IllegalArgumentException {
+    final String[][] values = new String[fileKeys.length][partitionColumnNames.length];
+    for (int fileIndex = 0; fileIndex < fileKeys.length; fileIndex++) {
+
+      for (int partitionIndex = 0; partitionIndex < partitionColumnNames.length; partitionIndex++) {
+        final Matcher matcher = Pattern.compile(partitionColumnNames[partitionIndex] + "=([^\\/\\\\]+)")
+                .matcher(fileKeys[fileIndex].toString());
+        if (!matcher.find()) {
+          throw new IllegalArgumentException(String.format("Unable to find partition column '%s' in file key '%s'",
```

Comment:
This is partially checked in ParseSetup phase to fail early and shot nice error message to the user, yet if someone would instruct the ParseHandler wrong programatically, it should also fail gracefully on this place.